idp/idputil/mfa/webauthn.go

```diff
@@ -0,0 +1,407 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package mfa contains utility routines for multi-factor
+// authentication with WebAuthn.
+package mfa
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/duo-labs/webauthn/protocol"
+	"github.com/duo-labs/webauthn/webauthn"
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/httprequest.v1"
+
+	"github.com/canonical/candid/idp/idputil"
+	"github.com/canonical/candid/idp/idputil/secret"
+	"github.com/canonical/candid/store"
+)
+
+var logger = loggo.GetLogger("candid.idp.idputil.mfa")
+
+type webauthnUser struct {
+	*store.Identity
+	Credentials []store.MFACredential
+}
+
+// WebAuthnID implements the webauthn.User interface.
+func (u *webauthnUser) WebAuthnID() []byte {
+	return []byte(u.Identity.ID)
```

Comment:
I suspect ProviderID would be better for this. The ID field is an internal-only idea. Theoretically it can change if we do a database migration, for instance.

Reply:
yes.. changed